File: backend-spring-boot/src/test/java/no/vicx/backend/user/service/RecaptchaServiceTest.java
package no.vicx.backend.user.service;

import no.vicx.backend.config.RestClientConfig;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.client.RestClientTest;
import org.springframework.context.annotation.Import;
import org.springframework.http.MediaType;
import org.springframework.test.web.client.MockRestServiceServer;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;

@RestClientTest(RecaptchaService.class)
@Import(RestClientConfig.class)
class RecaptchaServiceTest {

    @Autowired
    private RecaptchaService sut;

    @Autowired
    private MockRestServiceServer mockServer;

    @Test
    void verifyToken_givenValidToken_expectTrue() {
        mockServer.expect(requestTo(EXPECTED_URL))
                .andRespond(withSuccess(validResponseBody, MediaType.APPLICATION_JSON));

        assertTrue(sut.verifyToken("some-token"));
    }

    @Test
    void verifyToken_givenErrorResponse_expectFalse() {
        mockServer.expect(requestTo(EXPECTED_URL))
                .andRespond(withSuccess(errorResponseBody, MediaType.APPLICATION_JSON));

        assertFalse(sut.verifyToken("some-token"));
    }

    @ParameterizedTest
    @ValueSource(strings = {"", "{}"})
    void verifyToken_givenEmptyResponseBody_expectFalse(String body) {
        mockServer.expect(requestTo(EXPECTED_URL))
                .andRespond(withSuccess(body, MediaType.APPLICATION_JSON));

        assertFalse(sut.verifyToken("some-token"));
    }

    private static final String EXPECTED_URL =
            "https://www.google.com/recaptcha/api/siteverify?secret=6LeIxAcTAAAAAGG-vFI1TnRWxMZNFuojJ4WifJWe&response=some-token";

    private static final String validResponseBody = """
            {
                "success": true,
                "challenge_ts": "~challenge_ts~",
                "hostname": "~hostname~",
                "error-codes": []
            }
            """;

    private static final String errorResponseBody = """
            {
                "success": false,
                "challenge_ts": "~challenge_ts~",
                "hostname": "~hostname~",
                "error-codes": ["Some error"]
            }
            """;
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/user/service/RecaptchaServiceTest.kt
package no.vicx.backend.user.service

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.data.Row2
import io.kotest.data.forAll
import io.kotest.matchers.shouldBe
import no.vicx.backend.user.service.RecaptchaService.Companion.RECAPTCHA_VERIFY_BASE_URL
import no.vicx.backend.user.service.RecaptchaService.Companion.SECRET_REQUEST_PARAMETER
import no.vicx.backend.user.service.RecaptchaService.Companion.SITE_VERIFY_PATH
import no.vicx.backend.user.service.RecaptchaService.Companion.TOKEN_RESPONSE_PARAMETER
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.restclient.test.autoconfigure.RestClientTest
import org.springframework.http.MediaType
import org.springframework.test.web.client.MockRestServiceServer
import org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo
import org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess
import org.springframework.web.util.UriComponentsBuilder

@RestClientTest(RecaptchaService::class)
class RecaptchaServiceTest(
    mockServer: MockRestServiceServer,
    sut: RecaptchaService,
    @Value($$"${recaptcha.secret}") recaptchaSecret: String,
) : BehaviorSpec({

        val expectedUrl =
            UriComponentsBuilder
                .fromUriString("$RECAPTCHA_VERIFY_BASE_URL$SITE_VERIFY_PATH")
                .queryParam(SECRET_REQUEST_PARAMETER, recaptchaSecret)
                .queryParam(TOKEN_RESPONSE_PARAMETER, TOKEN_IN_TEST)
                .build()
                .toUri()

        Given("request with valid reCAPTCHA token") {
            mockServer
                .expect(requestTo(expectedUrl))
                .andRespond(
                    withSuccess(validResponseBody, MediaType.APPLICATION_JSON),
                )

            When("calling verifyToken") {
                val result = sut.verifyToken(TOKEN_IN_TEST)

                Then("expect true") {
                    result shouldBe true
                }
            }
        }

        Given("error response") {
            mockServer
                .expect(requestTo(expectedUrl))
                .andRespond(
                    withSuccess(errorResponseBody, MediaType.APPLICATION_JSON),
                )

            When("calling verifyToken") {
                val result = sut.verifyToken(TOKEN_IN_TEST)

                Then("expect false") {
                    result shouldBe false
                }
            }
        }

        Given("partial responses") {
            forAll(
                Row2("Empty body", ""),
                Row2("Empty JSON object", "{}"),
                // Funker ikke lenger med SB 4 Row2("JSON object with null field", "{\"success\": null}"),
            ) { description, body ->
                When("calling verifyToken: $description") {
                    mockServer
                        .expect(requestTo(expectedUrl))
                        .andRespond(
                            withSuccess(body, MediaType.APPLICATION_JSON),
                        )

                    val result = sut.verifyToken(TOKEN_IN_TEST)

                    Then("expect false") {
                        result shouldBe false
                    }
                }
            }
        }
    }) {
    companion object {
        private const val TOKEN_IN_TEST = "~token~"

        private val validResponseBody =
            """
            {
                "success": true,
                "challenge_ts": "~challenge_ts~",
                "hostname": "~hostname~",
                "error-codes": []
            }
            """.trimIndent()

        private val errorResponseBody =
            """
            {
                "success": false,
                "challenge_ts": "~challenge_ts~",
                "hostname": "~hostname~",
                "error-codes": ["Some error"]
            }
            """.trimIndent()
    }
}
